sel with method 'nearest' fails with AssertionError

The following fails

```python
def test_sel_nearest():
    """Test `sel` with method='nearest'"""
    # create test data
    times = pd.date_range('2000-01-01', periods=4)
    foo = xr.DataArray([1,2,3,4], coords=dict(time=times), dims='time')
    
    # works
    expected = foo.sel(time='2000-01-01')
    
    # fails
    assert foo.sel(time='2000-01-01', method='nearest') == expected
```

with an `AssertionError` in [`xarray/core/variable.py`](https://github.com/pydata/xarray/blob/v0.8.2/xarray/core/variable.py#L379):

```
C:\Users\uuu\AppData\Local\Continuum\Miniconda2\lib\site-packages\xarray\core\dataarray.pyc in sel(self, method, tolerance, **indexers)
    623             self, indexers, method=method, tolerance=tolerance
    624         )
--> 625         return self.isel(**pos_indexers)._replace_indexes(new_indexes)
    626 
    627     def isel_points(self, dim='points', **indexers):

C:\Users\uuu\AppData\Local\Continuum\Miniconda2\lib\site-packages\xarray\core\dataarray.pyc in isel(self, **indexers)
    608         DataArray.sel
    609         """
--> 610         ds = self._to_temp_dataset().isel(**indexers)
    611         return self._from_temp_dataset(ds)
    612 

C:\Users\uuu\AppData\Local\Continuum\Miniconda2\lib\site-packages\xarray\core\dataset.pyc in isel(self, **indexers)
    910         for name, var in iteritems(self._variables):
    911             var_indexers = dict((k, v) for k, v in indexers if k in var.dims)
--> 912             variables[name] = var.isel(**var_indexers)
    913         return self._replace_vars_and_dims(variables)
    914 

C:\Users\uuu\AppData\Local\Continuum\Miniconda2\lib\site-packages\xarray\core\variable.pyc in isel(self, **indexers)
    539             if dim in indexers:
    540                 key[i] = indexers[dim]
--> 541         return self[tuple(key)]
    542 
    543     def _shift_one_dim(self, dim, count):

C:\Users\uuu\AppData\Local\Continuum\Miniconda2\lib\site-packages\xarray\core\variable.pyc in __getitem__(self, key)
    377         # orthogonal indexing should ensure the dimensionality is consistent
    378         if hasattr(values, 'ndim'):
--> 379             assert values.ndim == len(dims), (values.ndim, len(dims))
    380         else:
    381             assert len(dims) == 0, len(dims)

AssertionError: (0, 1)
```

It does not matter which type the dimension has that is indexed:

```python
def test_sel_nearest_int():
    """Test `sel` with method='nearest'"""
    bar = xr.DataArray([1, 2, 3, 4], coords=dict(dummy=range(4)), dims='dummy')
    
    # works
    expected = bar.sel(dummy=3)
    
    # fails
    assert bar.sel(dummy=3, method='nearest') == expected
```

This is on Miniconda for Windows 64 bit with conda-forge and IOOS builds and
* xarray=0.8.2
* pandas=0.19.1
* numpy=1.11.2

Why might this be? Am I doing something wrong?

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -12,6 +12,11 @@ environment:
       PYTHON_ARCH: "32"
       CONDA_ENV: "py27-windows"
 
+    - PYTHON: "C:\\Python27-conda64"
+      PYTHON_VERSION: "2.7"
+      PYTHON_ARCH: "64"
+      CONDA_ENV: "py27-windows"
+
     - PYTHON: "C:\\Python36-conda64"
       PYTHON_VERSION: "3.6"
       PYTHON_ARCH: "64"

--- a/doc/whats-new.rst
+++ b/doc/whats-new.rst
@@ -40,6 +40,9 @@ Bug fixes
 - ``rolling`` now keeps its original dimension order (:issue:`1125`).
   By `Keisuke Fujii <https://github.com/fujiisoup>`_.
 
+- Fix ``sel`` with ``method='nearest'`` on Python 2.7 and 64-bit Windows
+  (:issue:`1140`).
+  `Stephan Hoyer <https://github.com/shoyer>`_.
 
 .. _whats-new.0.9.1:
 

--- a/xarray/backends/pydap_.py
+++ b/xarray/backends/pydap_.py
@@ -6,7 +6,7 @@ import numpy as np
 from .. import Variable
 from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin
 from ..core import indexing
-from ..pycompat import integer_types
+from ..core.pycompat import integer_types
 
 from .common import AbstractDataStore, robust_getitem
 

--- a/xarray/backends/pydap_.py
+++ b/xarray/backends/pydap_.py
@@ -6,6 +6,7 @@ import numpy as np
 from .. import Variable
 from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin
 from ..core import indexing
+from ..pycompat import integer_types
 
 from .common import AbstractDataStore, robust_getitem
 
@@ -29,7 +30,7 @@ class PydapArrayWrapper(NDArrayMixin):
         if not isinstance(key, tuple):
             key = (key,)
         for k in key:
-            if not (isinstance(k, (int, np.integer, slice)) or k is Ellipsis):
+            if not (isinstance(k, integer_types + (slice,)) or k is Ellipsis):
                 raise IndexError('pydap only supports indexing with int, '
                                  'slice and Ellipsis objects')
         # pull the data from the array attribute if possible, to avoid
@@ -38,7 +39,7 @@ class PydapArrayWrapper(NDArrayMixin):
         result = robust_getitem(array, key, catch=ValueError)
         # pydap doesn't squeeze axes automatically like numpy
         axis = tuple(n for n, k in enumerate(key)
-                     if isinstance(k, (int, np.integer)))
+                     if isinstance(k, integer_types))
         result = np.squeeze(result, axis)
         return result
 

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -27,7 +27,7 @@ from .utils import (Frozen, SortedKeysDict, maybe_wrap_array, hashable,
 from .variable import (Variable, as_variable, IndexVariable,
                        broadcast_variables)
 from .pycompat import (iteritems, basestring, OrderedDict,
-                       dask_array_type, range)
+                       integer_types, dask_array_type, range)
 from .options import OPTIONS
 
 # list of attributes of pd.DatetimeIndex that are ndarrays of time info
@@ -1107,7 +1107,7 @@ class Dataset(Mapping, ImplementsDatasetReduce, BaseDataObject,
 
         # all indexers should be int, slice or np.ndarrays
         indexers = [(k, (np.asarray(v)
-                         if not isinstance(v, (int, np.integer, slice))
+                         if not isinstance(v, integer_types + (slice,))
                          else v))
                     for k, v in iteritems(indexers)]
 

--- a/xarray/core/groupby.py
+++ b/xarray/core/groupby.py
@@ -11,7 +11,7 @@ from .combine import concat
 from .common import (
     ImplementsArrayReduce, ImplementsDatasetReduce, _maybe_promote,
 )
-from .pycompat import range, zip
+from .pycompat import range, zip, integer_types
 from .utils import hashable, peek_at, maybe_wrap_array, safe_cast_to_index
 from .variable import as_variable, Variable, IndexVariable
 
@@ -408,7 +408,7 @@ class GroupBy(object):
         return self._where(cond)
 
     def _first_or_last(self, op, skipna, keep_attrs):
-        if isinstance(self._group_indices[0], (int, np.integer)):
+        if isinstance(self._group_indices[0], integer_types):
             # NB. this is currently only used for reductions along an existing
             # dimension
             return self._obj

--- a/xarray/core/indexing.py
+++ b/xarray/core/indexing.py
@@ -7,7 +7,8 @@ import numpy as np
 import pandas as pd
 
 from . import utils
-from .pycompat import iteritems, range, dask_array_type, suppress
+from .pycompat import (iteritems, range, integer_types, dask_array_type,
+                       suppress)
 from .utils import is_full_slice, is_dict_like
 
 
@@ -81,7 +82,8 @@ def orthogonal_indexer(key, shape):
     # note: we treat integers separately (instead of turning them into 1d
     # arrays) because integers (and only integers) collapse axes when used with
     # __getitem__
-    non_int_keys = [n for n, k in enumerate(key) if not isinstance(k, (int, np.integer))]
+    non_int_keys = [n for n, k in enumerate(key)
+                    if not isinstance(k, integer_types)]
 
     def full_slices_unselected(n_list):
         def all_full_slices(key_index):
@@ -316,7 +318,7 @@ def slice_slice(old_slice, applied_slice, size):
 
 
 def _index_indexer_1d(old_indexer, applied_indexer, size):
-    assert isinstance(applied_indexer, (int, np.integer, slice, np.ndarray))
+    assert isinstance(applied_indexer, integer_types + (slice, np.ndarray))
     if isinstance(applied_indexer, slice) and applied_indexer == slice(None):
         # shortcut for the usual case
         return old_indexer
@@ -352,7 +354,7 @@ class LazilyIndexedArray(utils.NDArrayMixin):
         new_key = iter(canonicalize_indexer(new_key, self.ndim))
         key = []
         for size, k in zip(self.array.shape, self.key):
-            if isinstance(k, (int, np.integer)):
+            if isinstance(k, integer_types):
                 key.append(k)
             else:
                 key.append(_index_indexer_1d(k, next(new_key), size))
@@ -449,7 +451,7 @@ class NumpyIndexingAdapter(utils.NDArrayMixin):
 
     def _convert_key(self, key):
         key = expanded_indexer(key, self.ndim)
-        if any(not isinstance(k, (int, np.integer, slice)) for k in key):
+        if any(not isinstance(k, integer_types + (slice,)) for k in key):
             # key would trigger fancy indexing
             key = orthogonal_indexer(key, self.shape)
         return key
@@ -480,7 +482,7 @@ class DaskIndexingAdapter(utils.NDArrayMixin):
 
     def __getitem__(self, key):
         key = expanded_indexer(key, self.ndim)
-        if any(not isinstance(k, (int, np.integer, slice)) for k in key):
+        if any(not isinstance(k, integer_types + (slice,)) for k in key):
             value = self.array
             for axis, subkey in reversed(list(enumerate(key))):
                 value = value[(slice(None),) * axis + (subkey,)]

--- a/xarray/core/pycompat.py
+++ b/xarray/core/pycompat.py
@@ -3,6 +3,8 @@ from __future__ import division
 from __future__ import print_function
 import sys
 
+import numpy as np
+
 PY2 = sys.version_info[0] < 3
 PY3 = sys.version_info[0] >= 3
 
@@ -10,6 +12,7 @@ if PY3:  # pragma: no cover
     basestring = str
     unicode_type = str
     bytes_type = bytes
+    integer_types = (int, np.integer)
 
     def iteritems(d):
         return iter(d.items())
@@ -24,9 +27,10 @@ if PY3:  # pragma: no cover
     from urllib.request import urlretrieve
 else:  # pragma: no cover
     # Python 2
-    basestring = basestring
-    unicode_type = unicode
+    basestring = basestring  # noqa
+    unicode_type = unicode  # noqa
     bytes_type = str
+    integer_types = (int, long, np.integer)  # noqa
 
     def iteritems(d):
         return d.iteritems()

--- a/xarray/core/variable.py
+++ b/xarray/core/variable.py
@@ -15,7 +15,8 @@ from . import indexing
 from . import ops
 from . import utils
 from . import nputils
-from .pycompat import basestring, OrderedDict, zip, dask_array_type
+from .pycompat import (basestring, OrderedDict, zip, integer_types,
+                       dask_array_type)
 from .indexing import (PandasIndexAdapter, orthogonally_indexable)
 
 import xarray as xr  # only for Dataset and DataArray
@@ -373,7 +374,7 @@ class Variable(common.AbstractArray, utils.NdimSizeLenMixin):
         key = self._item_key_to_tuple(key)
         key = indexing.expanded_indexer(key, self.ndim)
         dims = tuple(dim for k, dim in zip(key, self.dims)
-                     if not isinstance(k, (int, np.integer)))
+                     if not isinstance(k, integer_types))
         values = self._indexable_data[key]
         # orthogonal indexing should ensure the dimensionality is consistent
         if hasattr(values, 'ndim'):

--- a/xarray/tests/test_dataset.py
+++ b/xarray/tests/test_dataset.py
@@ -23,7 +23,8 @@ import pytest
 from xarray import (align, broadcast, backends, Dataset, DataArray, Variable,
                     IndexVariable, open_dataset, set_options, MergeError)
 from xarray.core import indexing, utils
-from xarray.core.pycompat import iteritems, OrderedDict, unicode_type
+from xarray.core.pycompat import (iteritems, OrderedDict, unicode_type,
+                                  integer_types)
 from xarray.core.common import full_like
 
 from . import (TestCase, InaccessibleArray, UnexpectedDataAccess,
@@ -2797,7 +2798,7 @@ class TestDataset(TestCase):
     def test_reduce_only_one_axis(self):
 
         def mean_only_one_axis(x, axis):
-            if not isinstance(axis, (int, np.integer)):
+            if not isinstance(axis, integer_types):
                 raise TypeError('non-integer axis')
             return x.mean(axis)
 

